### ISSUE TITLE
master branch kube's private internal subnets route table uses igw instead of NAT

### ISSUE THREAD

**Issue by cmcconnell1** [kubernetes-incubator/kube-aws#290]
Hello,

As of the pre-release version 0.9.3 and building kube-aws binary from forked master branch, it seems as though the networking/routes for using private, internal NAT subnets as configured in cluster.yaml for kube-aws are no longer working for us.
We are using the all of the same values (in our cluster.yaml) that we've been using successfully--up until this pre-release version 0.9.3, including:
- vpcId
- routeTableId
- vpcCIDR
- subnets and their associated instanceCIDR ranges
- serviceCIDR
- podCID

Note: Some of the content for this has come from my below two comments from one of our existing issues:
- https://github.com/coreos/kube-aws/issues/189#issuecomment-275816250
- https://github.com/coreos/kube-aws/issues/189#issuecomment-276199562

Route tables from kube-aws-v0.9.3-rc.5 kubernetes subnets correctly generated and sent traffic to our NAT instance:
```
# v0.9.3-rc.5 myapp-kube-Subnet0 sends all traffic to EC2 NAT instance
 0.0.0.0/0     eni-def0049a / i-abc12345

# v0.9.3-rc.5 myapp-kube-Subnet1 sends all traffic to EC2 NAT instance
 0.0.0.0/0     eni-def0049a / i-abc12345
```

Note that this is correct and works for us and we confirm that this matches our other existing internal non-kube-aws/kubernetes subnets as shown below (wherein the route table sends all traffic to our EC2 NAT instance).
```
0.0.0.0/0     eni-def0049a / i-abc12345
```

Now with the latest pre-release 0.9.3 version of kube-aws, what we see (with us using the same settings and configurations that we've been using for all releases up to and including v0.9.3-rc.5) is that all of the kube-aws provisioned kubernetes nodes cannot access the internet. Note that many of the specific and supporting details are documented in the above two comments/links.

What appears to be the problem for us, is that now kube-aws is configuring the kubernetes subnets to try to send traffic to either a specified existing, or newly created internet gateway, instead of forwarding the private subnet traffic to (our EC2 instance which does the) NAT.

```
# pre-release 0.9.3 myapp-kube-Subnet0 route table sends traffic to an igw instead of the NAT instance
0.0.0.0/0     igw-ccb87ca4 

# pre-release 0.9.3 myapp-kube-Subnet0 route table sends traffic to an igw instead of the NAT instance
0.0.0.0/0     igw-ccb87ca4 
```

Also from the new pre-release kube-aws version's generated cluster.yaml file, noting that we now see config options for specifying an existing ```natGateway``` wherein it appears as though **perhaps a very important distinction should/must be made that this new configuration option**, and its related code, appear to be designed to work with the newer _**AWS NAT Gateway service**_ (and **perhaps/maybe not** with the (older) _**EC2 NAT instances**?_  Because if it did support actual EC2 NAT instances,  we'd need to specify an actual EC2 instance-id of the current NAT instance?  I don't see this noted in the comments and **perhaps I am in error by not noting this distinction** when I tested building and deploying the pre-release binary?

```
# Kubernetes subnets with their CIDRs and availability zones. Differentiating availability zone for 2 or more subnets result in high-availability (failures of a single availability zone won't result in immediate downtimes)
# subnets:
#   - availabilityZone: us-west-1a
#     instanceCIDR: "10.0.0.0/24"
#     subnetId: "subnet-xxxxxxxx" #optional
#   - availabilityZone: us-west-1b
#     instanceCIDR: "10.0.1.0/24"
#     # natGateway:
#     #   # Pre-allocated NAT Gateway. Used with private subnets.
#     #   id: "ngw-abcdef12"
#     #   # Pre-allocated EIP for NAT Gateways. Used with private subnets.
#     #   eipAllocationId: "eipalloc-abcdef12"
#     subnetId: "subnet-xxxxxxxx" #optional
```

To summarize:
The kube-aws provisioned kubernetes subnet routing has changed for us between the latest (and last working version for us) v0.9.3-rc.5 and the current state of the master branch with pre-release 0.9.3
I'm confused as to why we would be forced into specifying (and if not, having kube-aws auto generating) an igw and then have kube-aws configure the routing of all traffic to it, if we are deploying into a private subnet which utilizes (and has been working correctly with previous kube-aws versions) NAT?

A few questions that would greatly help us move forward:
Are the new configuration options (and related/supported code) _only_ applicable for the (perhaps newer) **AWS NAT Gateway service** and **NOT** applicable for (older) infrastructures which may utilize an **actual NAT EC2 instance?**  If so, this will now be a problem for us as.

If the changes are by design, then with the latest pre-release and onward, we will need to know how to correctly configure our cluster.yaml per our requirements, including: 
- deploying into existing VPC with existing igw.
- deploying into private subnets which **use an actual EC2 NAT instance and not the NAT Service for all NAT traffic** (ensuring the requisite route table is specified in cluster.yaml) 
- want multi-AZ H/A

Thank you!

**Comment by mumoshu** [comment]
Hi @cmcconnell1, thanks as always!

I'm running short on time but AFAICS right now, the current master branch is for deverlopers only; it isn't a v0.9.3 candidate but a v0.9.4-rc.1 candidate.

v0.9.3 is still pending due to the binary signing process but you can build it your own by checking out the `v0.9.3` tag of the git repo.

Let me also add that I believe the bug you've mentioned will be fixed once #284 is merged.

**Comment by mumoshu** [comment]
> A few questions that would greatly help us move forward:
Are the new configuration options (and related/supported code) only applicable for the (perhaps newer) AWS NAT Gateway service and NOT applicable for (older) infrastructures which may utilize an actual NAT EC2 instance? If so, this will now be a problem for us as.

No!
Probably you're trying an incomplete version of kube-aws built from pre-v0.9.4-r.c1 but various issues in it were fixed via #284.
As noted in my prev comment, the current master branch isn't pre-v0.9.3 but pre-v0.9.4-rc.1.

Specifically to your use-case, please try with the latest v0.9.4-rc.1 released just now and provide kube-aws your own route table with a route to your own NAT already added to it, by specifying `subnets[].routeTable.id` in `cluster.yaml`.
You may already know but please let me extra clear that it's an user's responsibility to configure an existing route table appropriately.

Also note that I've added many comments to `cluster.yaml` to describe intended-to-be-supported network topology via #284.

**Comment by cmcconnell1** [comment]
Thanks @mumoshu for your notes both here and offline.  I was indeed building  from the master branch and did not pull the specific version's tag.  I changed the title to reflect this.

On a somewhat related note, I wanted to provide some of my notes/observations for the latest kube-aws **v0.9.4-rc.1** somewhere, so instead of creating another issue, I'll just post them here.

Unfortunately, due to time constraints, I was not able to continue working with the latest 0.9.4-rc.1 version for very long.  I spent most of one day with it and loved what I saw with the new features, etc.   I tried deploying into both internal (private EC2 subnets with existing and working NAT route tables) and external (with working IGW route tables) subnet topologies, but both seemed to have routing issues **_for us_** (again this could be just for us with our existing VPC and its configurations which may be problematic at times).  

On a good note, now seeing these as new in CF Outputs and this should hopefully address our issues as noted in https://github.com/coreos/kube-aws/issues/189
```
InstanceEtcd1PrivateDnsName ip-10-1-11-175.us-west-1.compute.internal The resolvable hostname of etcd node 1  clustername-InstanceEtcd1PrivateDnsName
InstanceEtcd2PrivateDnsName ip-10-1-10-193.us-west-1.compute.internal The resolvable hostname of etcd node 2  clustername-InstanceEtcd2PrivateDnsName
InstanceEtcd0PrivateDnsName ip-10-1-10-9.us-west-1.compute.internal The resolvable hostname of etcd node 0  clustername-InstanceEtcd0PrivateDnsNam

```
Some additional notes for 0.9.4-rc.1 which I'm not sure is what we'd want:

Looking at a fresh kube-aws provisioned (internal private NAT subnet) cluster:
kube-aws created a NAT Gateway (via a service) for its subnet (with name tag: clustername-FooKubePrivate1a) _even though we specified an existing route table ID_ for all subnets which had a route table that utilized an EC2 NAT instance for NAT.  

kube-aws created the subnet which had the new following new route table:  
```rtb-75ad1311 | clustername-FooKubePrivate1aRouteTable```
and sent traffic to a new NAT gateway ID nat-0367bec78c074f7a0 which had an EIP.  

The above configuration was not desired (or in our case functional); I think we'd expect it to use the existing specified route table with our existing EC2 NAT instance?

Since there were no examples of exactly how to us the new syntax of ```subnets[].routeTable.id``` I tried using the below, and hope this is correct?:
```
subnets:
  # INTERNAL private NAT subnet in AZ-1a
  - name: foo-private-1a
    # Set to false if this subnet is public
    private: true
    availabilityZone: us-west-1a
    instanceCIDR: "10.1.10.0/24"
    subnets[].routeTable.id: "rtb-acb12345"  # internal NAT
```

and for etcd configuration directive specifically:
```
etcd:
  # If omitted, public subnets are created by kube-aws and used for etcd nodes
  subnets:
    # References subnets defined under the top-level `subnets` key by their names
    #- name: ManagedPrivateSubnet1
    - name: foo-private-1a
    #- name: ManagedPrivateSubnet2
    - name: foo-private-1b 
```

Unfortunately, the etcd nodes were not able to connect to the internet.  The controller was though, but since the etcd nodes were not able to connect, they were not provisioned correctly and we have the domino effect with our kubernetes cluster provisioning failing. 

Since I had a deliverable (**and kube-aws v0.9.3-rc.5 works fine for us**),  I punted on the latest 0.9.4-rc.1 version, and downgraded back to rc.5.  

With a fresh rc.5 kube-aws deploy ( and using our previously noted cloud-init hostname work-around as documented in: https://github.com/coreos/kube-aws/issues/189 ) we were again immediately able to deploy working kubernetes clusters and configure and deploy working H/A node pools, etc.

Thanks as always

**Comment by mumoshu** [comment]
@cmcconnell1 Thanks for the detailed feedback as always!

Just a quick reply but the `subnets` should be:

```yaml
subnets:
  # INTERNAL private NAT subnet in AZ-1a
  - name: foo-private-1a
    # Set to false if this subnet is public
    private: true
    availabilityZone: us-west-1a
    instanceCIDR: "10.1.10.0/24"
    routeTable:
      id: "rtb-acb12345" # internal NAT
    # instead of:
    #subnets[].routeTable.id: "rtb-acb12345"  # internal NAT
```

**Comment by cmcconnell1** [comment]
Thanks @mumoshu, suppose my bad here, but would be great if we could get examples of how to use any new configuration options (perhaps commented-out, etc.) in the cluster.yaml file?

I was wondering why I had that wrong and just checked a new default cluster.yaml file created with the kube-aws version 0.9.4-rc.1 and we see this: 

```
# More concretely,
# * If you like to make all the subnets private, pre-configure an NGW yourself and add a route to the NGW beforehand
# * If you like to make all the subnets public, pre-configure an IGW yourself and add a route to the IGW beforehand
# * If you like to mix private and public subnets, omit routeTableId but specify subnets[].routeTable.id per subnet
#
# routeTableId: rtb-xxxxxxxx
```
Appreciate your comment above but then this commented out config/code from a new cluster.yaml  doesn't appear correct then does it?  Specifically the ```routeTableId: rtb-xxxxxxxx``` directive?  

It would be great if we could have your previous comment/example above added as default to new cluster.yaml files (perhaps just commented-out?) so folks can see where and how to use any new syntax correctly.  
I'll admit that we saw a lot of changes going from rc.5 to v0.9.4-rc.1 and with all those changes (and the default commented-out code) I seemed to miss using this one correctly.

I hope to get some free time to re-test the latest version again

Thanks!

**Comment by mumoshu** [comment]
@cmcconnell1 Could you confirm that this is fixed in the latest rc? Thanks!

**Comment by mumoshu** [comment]
Note: Confirmed to work with v0.9.6-rc.2 according to https://github.com/kubernetes-incubator/kube-aws/issues/189#issuecomment-294894954

Thanks for your confirmation @cmcconnell1 👍
